### PR TITLE
[FSSDK-9562] refactor odp event api manager to handle odp config

### PR DESCRIPTION
**Pull request by raju-opti**
## Summary
- Refactor OdpEventApiManager to handle odp config update.

## Test plan
- Refactored affected existing tests
- Added new tests for config udpates

## Issues
- FSSDK-9562

### CODE REVIEW

File: packages/optimizely-sdk/tests/odpEventManager.spec.ts [hunk 1]
```diff
@@ -155,6 +153,55 @@ describe('OdpEventManager', () => {
     resetCalls(mockApiManager);
   });
 
+  it('should update api manager setting with odp config on instantiation', () => {
+    when(mockApiManager.sendEvents(anything())).thenResolve(false);
+    when(mockApiManager.updateSettings(anything())).thenReturn(undefined);
+
+    const apiManager = instance(mockApiManager); 
+
+    const eventManager = new OdpEventManager({
```
Comment (mikechu-optimizely) [inline packages/optimizely-sdk/tests/odpEventManager.spec.ts:162]: Would it be possible to use the discard operator? Curious.

File: packages/optimizely-sdk/tests/odpEventManager.spec.ts [hunk 2]
```diff
@@ -378,77 +423,75 @@ describe('OdpEventManager', () => {
...
+    const apiManager = instance(mockApiManager);
+
     const eventManager = new OdpEventManager({
       odpConfig,
       apiManager,
       logger,
       clientEngine,
       clientVersion,
-      flushInterval: 100,
+      batchSize: 10, // with batch size of 10...
```
Comment (mikechu-optimizely) [inline packages/optimizely-sdk/tests/odpEventManager.spec.ts:473]: Yeah I use comments too much.

**Review by mikechu-optimizely — CHANGES_REQUESTED** [review]
Some thoughts.

File: packages/optimizely-sdk/lib/plugins/odp/event_api_manager/index.browser.ts [hunk 3]
```diff
@@ -15,10 +16,15 @@ export class BrowserOdpEventApiManager extends OdpEventApiManager {
...
-    apiKey: string,
     events: OdpEvent[]
   ): { method: string; endpoint: string; headers: { [key: string]: string }; data: string } {
+    // the caller should ensure odpConfig is ready before calling
+    if (!this.odpConfig?.isReady()) {
+      this.getLogger().log(LogLevel.ERROR, ODP_CONFIG_NOT_READY_MESSAGE);
+      throw new Error(ODP_CONFIG_NOT_READY_MESSAGE);
+    }
+
+    const apiKey = this.odpConfig!.apiKey;
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply packages/optimizely-sdk/lib/plugins/odp/event_api_manager/index.browser.ts:27]: Yep. non-null assertion is not needed. The previous if clause ensures that the value is present and typescript can narrow the type. I'll remove the assertion.

File: packages/optimizely-sdk/lib/core/odp/odp_event_api_manager.ts [hunk 4]
```diff
@@ -17,14 +17,18 @@
...
 
 const EVENT_SENDING_FAILURE_MESSAGE = 'ODP event send failed';
+export const ODP_CONFIG_NOT_READY_MESSAGE = 'ODP config not ready';
 
 /**
  * Manager for communicating with the Optimizely Data Platform REST API
  */
 export interface IOdpEventApiManager {
-  sendEvents(apiKey: string, apiHost: string, events: OdpEvent[]): Promise<boolean>;
+  // sendEvents(apiKey: string, apiHost: string, events: OdpEvent[]): Promise<boolean>;
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply packages/optimizely-sdk/lib/core/odp/odp_event_api_manager.ts:29]: sure

File: packages/optimizely-sdk/lib/core/odp/odp_event_api_manager.ts [hunk 5]
```diff
@@ -53,6 +62,14 @@ export abstract class OdpEventApiManager implements IOdpEventApiManager {
     this.logger = logger;
   }
 
+  /**
+   * Updatets odpConfig of the api manager instance
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply packages/optimizely-sdk/lib/core/odp/odp_event_api_manager.ts:66]: done

File: packages/optimizely-sdk/lib/core/odp/odp_event_api_manager.ts [hunk 6]
```diff
@@ -64,11 +81,11 @@ export abstract class OdpEventApiManager implements IOdpEventApiManager {
    * @param events ODP events to send
    * @returns Retry is true - if network or server error (5xx), otherwise false
    */
-  async sendEvents(apiKey: string, apiHost: string, events: OdpEvent[]): Promise<boolean> {
+  async sendEvents(events: OdpEvent[]): Promise<boolean> {
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply packages/optimizely-sdk/lib/core/odp/odp_event_api_manager.ts:84]: Fixed JSDoc.

 The interface is not part of the public API, this is just an internal implementation detail. Should be okay.

File: packages/optimizely-sdk/tests/odpEventApiManager.spec.ts [hunk 7]
```diff
@@ -68,7 +79,7 @@ describe('NodeOdpEventApiManager', () => {
     );
     const manager = managerInstance();
 
-    const shouldRetry = await manager.sendEvents(VALID_ODP_PUBLIC_KEY, ODP_REST_API_HOST, ODP_EVENTS);
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply packages/optimizely-sdk/tests/odpEventApiManager.spec.ts:71]: Yep

File: packages/optimizely-sdk/tests/odpEventApiManager.spec.ts [hunk 8]
```diff
@@ -105,9 +116,37 @@ describe('NodeOdpEventApiManager', () => {
...
+
+    await manager.sendEvents(ODP_EVENTS);
+
+    const updatetdOdpConfig = new OdpConfig(
+      'updated-key',
+      'https://updatedhost.test',
+      ['updated-seg'],
+    )
+    
+    manager.updateSettings(updatetdOdpConfig);
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply packages/optimizely-sdk/tests/odpEventApiManager.spec.ts:141]: yep, made that typo in couple of places :(

File: packages/optimizely-sdk/tests/odpEventManager.spec.ts [hunk 9]
```diff
@@ -155,6 +153,55 @@ describe('OdpEventManager', () => {
...
+
+    const eventManager = new OdpEventManager({
+      odpConfig,
+      apiManager,
+      logger,
+      clientEngine,
+      clientVersion,
+    });
+
+    const updatetdOdpConfig = new OdpConfig(
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply packages/optimizely-sdk/tests/odpEventManager.spec.ts:188]: aargh! fixing

File: packages/optimizely-sdk/tests/odpEventManager.spec.ts [hunk 10]
```diff
@@ -378,77 +423,75 @@ describe('OdpEventManager', () => {
...
+
     const eventManager = new OdpEventManager({
       odpConfig,
       apiManager,
       logger,
       clientEngine,
       clientVersion,
-      batchSize: 10,
-      flushInterval: 100,
+      batchSize: 10, // with batch size of 10...
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply packages/optimizely-sdk/tests/odpEventManager.spec.ts:437]: removing

File: packages/optimizely-sdk/tests/odpEventManager.spec.ts (hunk 2 above)
Comment (raju-opti) [reply packages/optimizely-sdk/tests/odpEventManager.spec.ts:473]: and I copy paste

File: packages/optimizely-sdk/tests/odpEventManager.spec.ts (hunk 1 above)
Comment (mikechu-optimizely) [reply packages/optimizely-sdk/tests/odpEventManager.spec.ts:162]: Thanks for clarifying that `_` is just another variable name in JS....my C# does not apply here 😄

**Review by mikechu-optimizely — APPROVED** [review]
LGTM

**Pull request merged**